### PR TITLE
Decaffeinate collection lists

### LINKED ISSUE #6515: Can't view Favorites from Project page, if not logged in
## Bug

If you're _not logged in_ and then go to (any Project) -> Collect -> Favorites -> (any Favorites), the page will crash with the error: `Cannot read properties of null (reading 'id')`

- Example URL: https://www.zooniverse.org/projects/penguintom79/penguin-watch/collections/mary2021/favorites-penguintom79-penguin-watch

Expectation:
- You should be able to view a Favorites/Favourites collection from the Project page _without having to log in._

Notes:
- If you access the Favorites from the Project page while you're _logged in,_ it works fine.
- If you access the Favorites from the _User_ page while you're _not logged in,_ it also works fine. Example: https://www.zooniverse.org/collections/mary2021/favorites-penguintom79-penguin-watch
- If you access a _Collection_ from the Project page while not logged in, it also also works fine. (???) Example: https://www.zooniverse.org/projects/penguintom79/penguin-watch/collections/chescamartin/08-09-21-2nd-week (How... how is this different?)

### Status

No volunteer has filed a complaint yet, and the affected functionality doesn't seem major. I'm marking this as a low-priority bug for now.

### PR DESCRIPTION
**Pull request by eatyourgreens**
- Refactor the collections `CollectionShowList` component as JSX.
- Remove `createReactClass`.
- Fix a crash when there is no logged-in user.

Staging branch URL: https://pr-6516.pfe-preview.zooniverse.org

Fixes #6515.

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `npm ci` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on GitHub Actions?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

File: app/collections/show-list.jsx [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+  selecting = false,
+  selected = false,
+  subject,
+  user = null
+}) {
+  const [isFavorite, setIsFavorite] = useState(false);
+  const [project, setProject] = useState(projectContext);
+
+  async function checkFavorite(project) {
+    if (collection?.favorite && (collection.links.owner.id === user?.id)) {
```
Comment (shaunanoordin) [inline app/collections/show-list.jsx:51]: (1/2) Ahhah, was this the culprit that caused #6515 ?

File: app/collections/show-list.cjsx [hunk 2]
```diff
@@ -1,290 +0,0 @@
...
-      @props.subject.get('project')
-        .then (project) =>
-          @updateProjectState(project)
-
-  updateProjectState: (project) ->
-    @setState {project}
-    @isFavorite(project)
-
-  isFavorite: (project) ->
-    if @props.collection.favorite and @props.collection.links.owner.id is @props.user.id
```
Comment (shaunanoordin) [inline app/collections/show-list.cjsx:49]: (2/2) Ahhah, I see it now. 👍

**Review by shaunanoordin — APPROVED** [review]
## PR Review

This PR updates the Collections-listing component, changing it from CoffeeScript to modern JS.

- Functionality is basically the same, as observed at...
  - Collections (e.g. [www.zooniverse.org/collections/mary2021/favorites-penguintom79-penguin-watch](https://pr-6516.pfe-preview.zooniverse.org/collections/mary2021/favorites-penguintom79-penguin-watch) and ...[/talk](https://pr-6516.pfe-preview.zooniverse.org/collections/mary2021/favorites-penguintom79-penguin-watch/talk))
  - Project -> Collections (e.g. [www.zooniverse.org/projects/penguintom79/penguin-watch/collections/mary2021/favorites-penguintom79-penguin-watch](https://pr-6516.pfe-preview.zooniverse.org/projects/penguintom79/penguin-watch/collections/mary2021/favorites-penguintom79-penguin-watch) and ...[/talk](https://pr-6516.pfe-preview.zooniverse.org/projects/penguintom79/penguin-watch/collections/mary2021/favorites-penguintom79-penguin-watch/talk))
    - Project (with locale) -> Collections (e.g. [www.zooniverse.org/projects/en/penguintom79/penguin-watch/collections/mary2021/favorites-penguintom79-penguin-watch](https://pr-6516.pfe-preview.zooniverse.org/projects/en/penguintom79/penguin-watch/collections/mary2021/favorites-penguintom79-penguin-watch) and ...[/talk](https://pr-6516.pfe-preview.zooniverse.org/projects/en/penguintom79/penguin-watch/collections/mary2021/favorites-penguintom79-penguin-watch/talk))
- Bug fix: Collections list displays whether the viewer is logged in or not. (Fixes #6515)

Changes make sense, and the fix seems to be a straightforward `user.id` -> `user?.id` on top of the automated decaffeniation.

### Testing

Tested with macOS + Chrome 111 using the staging URL, targeting a random project and a random user's favourites collection.

Testing steps:
- View all the links under "Functionality is basically the same" bit above, twice (1) when logged in, and (2) while anonymous.

All good 👌 

### Status

LGTM 👍

File: app/collections/show-list.jsx [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+      setProject(project);
+    }
+
+    if (projectContext) {
+      if (projectContext !== project) {
+        setProject(projectContext);
+      }
+    } else {
+      getSubjectProject();
+      checkFavorite(project);
```
Comment (eatyourgreens) [inline app/collections/show-list.jsx:81]: Should this run after the project has updated? Here, it's running before the project is fetched.

**Review by eatyourgreens — COMMENTED** [review]
`checkFavorite` should wait until the project changes. Derived state, like this, should really be in the main body of the function. 

https://react.dev/reference/react/useState#storing-information-from-previous-renders

**Comment by eatyourgreens** [comment]
I found a bug in my JS (introduced when I removed `createReactClass`.) Should be fixed now.

**Comment by eatyourgreens** [comment]
Selecting subjects works now. Also adding/removing them from collections, so I'm going to merge this. It might be worth double-checking on staging, though, before deploying this. It rewrote all of the collection management functions from coffeescript to JS.

**Pull request merged**